planner/util/sqlrestorer/query_block_print.go

```diff
@@ -0,0 +1,378 @@
+		first = false
+	}
+	return res
+}
+
+// ExprToString print an expression as a string.
+// useProjectedCol means if projected columns can be used when printing.
+// Example:
+//
+//	There is a projected column: a + 10 as n, and n's unique ID is 2.
+//	When trying to print a column with unique ID is 2, if useProjectedCol is true, it will be n,
+//	 otherwise, it will be a + 10
+func (q *QueryBlock) ExprToString(e expression.Expression, useProjectedCol bool) (string, error) {
+	switch expr := e.(type) {
+	case *expression.ScalarFunction:
+		var buffer bytes.Buffer
+		buffer.WriteString("`" + expr.FuncName.L + "`(")
+		for i, arg := range expr.GetArgs() {
+			argStr, err := q.ExprToString(arg, useProjectedCol)
+			if err != nil {
+				return "", err
+			}
+			buffer.WriteString(argStr)
+			if i+1 != len(expr.GetArgs()) {
+				buffer.WriteString(", ")
+			}
+		}
+		// TODO: CAST is not equal to the original expression, but this doesn't affect cardinality estimation.
+		buffer.WriteString(")")
+		return buffer.String(), nil
+	case *expression.Column:
+		uid := expr.UniqueID
+		name, ok := q.findAndMarkColName(uid, useProjectedCol)
+		if ok {
+			return name, nil
+		}
+		return unknownColumnPlaceholder + strconv.FormatInt(uid, 10), nil
+	case *expression.CorrelatedColumn:
+		uid := expr.Column.UniqueID
+		return unknownColumnPlaceholder + strconv.FormatInt(uid, 10), nil
+	case *expression.Constant:
+		value, err := expr.Eval(chunk.Row{})
+		if err != nil {
+			return "", err
+		}
+		valueExpr := driver.ValueExpr{Datum: value}
+		var buffer bytes.Buffer
+		restoreCtx := format.NewRestoreCtx(format.DefaultRestoreFlags, &buffer)
+		err = valueExpr.Restore(restoreCtx)
+		if err != nil {
+			return "", err
+		}
+		return buffer.String(), nil
+	}
+	return "", errors.New("unexpected type of Expression")
+}
+
+// ByItemToString print ByItems as a string.
+func (q *QueryBlock) ByItemToString(items []*util.ByItems, useProjectedCol bool) (string, error) {
+	str := ""
+	for i, item := range items {
+		if i != 0 {
+			str += ", "
+		}
+		s, err := q.ExprToString(item.Expr, useProjectedCol)
+		if err != nil {
+			return s, err
+		}
+		str += s
+		if item.Desc {
+			str += " DESC "
+		}
+	}
+	return str, nil
+}
+
+// AggFuncToString print an aggregate function as a string.
+func (q *QueryBlock) AggFuncToString(agg *aggregation.AggFuncDesc) (string, error) {
+	if agg.Name == "firstrow" {
+		// Projected columns can't be used in aggregate functions, so useProjectedCol is false.
+		s, err := q.ExprToString(agg.Args[0], false)
+		if err != nil {
+			return s, err
+		}
+		return s, nil
+	}
+	str := agg.Name + "("
+	first := true
+	for i := 0; i < len(agg.Args)-1; i++ {
+		if !first {
+			str += ", "
+			first = false
+		}
+		s, err := q.ExprToString(agg.Args[i], false)
+		if err != nil {
+			return s, err
+		}
+		str += s
+	}
+	if agg.Name == "group_concat" {
+		if agg.OrderByItems != nil {
+			str += " ORDER BY "
+			s, err := q.ByItemToString(agg.OrderByItems, false)
```

Comment:
seems this is the func's only usage, why bother adding a false param here?

Reply:
We may add support for `LogicalSort` in the future.
The `ORDER BY` clause in the SQL can use projected columns, so we just reserve this parameter in advance.